tagExpression/utils.js

```diff
@@ -1,23 +1,136 @@
 const _ = require('lodash');
-const { LexerParser } = require('./LexerParser');
+const { LexerParser, LexerToken } = require('./LexerParser');
+
+/**
+ * Helper method to convert Tag expressions with tag names having special chars to
+ * Lexer Parser safe format, and provide mapping back to tag names after parsing
+ * @param formula - current formula to be parsed
+ * @param tagNames - list of valid existing tags
+ *
+ * @returns {object} -> Contains LexParser safe tag expression and tagToUniqueID mapping
+ *
+ *  { 	
+ * 		originalFormula: string,
+ * 		lexSafeFormula: string,
+ * 		tagNameToUniqueIDMap:Map<string, string>,
+ * 		allTagsValid: boolean	
+ *  }
+ * tagNameToUniqueIDMap : key -> Tag Identifier is the special character trimmed in the original Tag
+ * tagNameToUniqueIDMap : value -> Mapped TagName
+ *
+ * @private
+ */
+function getLexParserSafeTagFormula(formula, tagNames) {
+
+    const assetTagRegex =
+        // eslint-disable-next-line no-useless-escape
+        /(?:(?:\$)(?:(?!\|).)*)(?:\|)((?:(?![\s\/\+\-\*\^\%\>\<\=\(\)\,][\$\&][\w].+?\|).)*\|)/gmi;
+    const listOfMatchedGroups = [];
+    while (null !== (matchTag = assetTagRegex.exec(formula))) {
+        listOfMatchedGroups.push(matchTag);
+    }
+    const tagMapToUniqueTagID = new Map();
+    let formulaForTheLexerParser = '';
+    let lastIndexReplaced = 0;
+
+    // eslint-disable-next-line no-useless-escape
+    const specialCharRegex = /[\#\$\.\,\+\?\|\&\-\_\=\>\^\@\(\)\[\]\\\/\'\%\!\*\`\{\}\~\<\"\:\;]/g;
+
+    const lengthOfMatchedGroups = listOfMatchedGroups.length;
+
+    let allTagsValid = true;
+    listOfMatchedGroups.every((currentValue, currentIndex) => {
```

Comment:
On the backend, do we need to check if any asset-tags matches present as well ?
i.e lengthOfMatchedGroups > 0

Reply:
Added the check